lib/createCookieEncryptor.js

```
@@ -0,0 +1,72 @@
+const base64url = require('base64url')
+const { BadRequestException } = require('./utils/errorHandlers')
+
+const toHexString = bytes =>
+  bytes.reduce((str, byte) => str + byte.toString(16).padStart(2, '0'), '')
+
+const fromHexString = hexString =>
+  new Uint8Array(hexString.match(/.{1,2}/g).map(byte => parseInt(byte, 16)))
+
+export const createCookieEncryptor = ({ cookieName, encryptionKey, keyAlgorithm }) => {
+  const getKey = async () => {
+    try {
+      return crypto.subtle.importKey(
+        'raw',
+        base64url.toBuffer(encryptionKey),
+        { name: keyAlgorithm },
+        false,
+        ['encrypt', 'decrypt']
+      )
+    } catch (e) {
+      throw BadRequestException(e.message || e.toString())
+    }
+  }
+
+  const encrypt = async str => {
+    try {
+      const key = await getKey(keyAlgorithm)
+      const iv = crypto.getRandomValues(new Uint8Array(16))
+      const encoder = new TextEncoder()
+      const data = encoder.encode(str)
+      const encrypted = await crypto.subtle.encrypt({ name: keyAlgorithm, iv }, key, data)
+      return `${toHexString(iv)}${base64url(encrypted)}`
+    } catch (e) {
+      throw BadRequestException(e.message || e.toString())
+    }
+  }
+
+  const decrypt = async str => {
+    try {
+      const iv = fromHexString(str.slice(0, 32))
+      const token = str.slice(32)
+      const key = await getKey(keyAlgorithm)
+      const decrypted = await crypto.subtle.decrypt({
+        name: keyAlgorithm, iv
+      }, key, base64url.toBuffer(token))
+      const decoder = new TextDecoder()
+      return decoder.decode(decrypted)
+    } catch (e) {
+      console.error({ decryptError: e })
```

Can we throw an error here as well?

---

whoops! missed that one